Update github.com/imdario/mergo calls from mergo.Merge to mergo.MergeWithOverwrite

bufdev opened the pull request:
Fixes https://github.com/GoogleCloudPlatform/kubernetes/issues/12272
Also per https://github.com/imdario/mergo/releases/tag/0.2.0

CLA signed and passes unit tests in `pkg/client/...`:

```
$ cd pkg/client && godep go test ./...
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client    0.353s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache  0.242s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/chaosclient    0.006s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd  0.027s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api  0.017s
?       github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api/latest   [no test files]
?       github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd/api/v1   [no test files]
?       github.com/GoogleCloudPlatform/kubernetes/pkg/client/metrics    [no test files]
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/portforward    0.016s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/record 0.057s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/remotecommand  0.034s
ok      github.com/GoogleCloudPlatform/kubernetes/pkg/client/testclient 0.012s
```

Commit:
Improve clarity around PodFitsResource by showing pods limits in `kubectl describe node`

bufdev committed:
Update use of mergo package to use mergo.MergeWithOverwrite instead of mergo.Merge

bufdev committed:
Update use of mergo package to use mergo.MergeWithOverwrite instead of mergo.Merge

bufdev committed:
Godep update github.com/imdario/mergo

bufdev committed:
Godep update github.com/imdario/mergo

peter-edge commented:
I could be wrong but I don't think the Shippable failure is real

dchen1107 commented:
ok to test

dchen1107 commented:
cc/ @deads2k

deads2k commented on pkg/client/clientcmd/loader.go:
```diff
@@ -134,8 +134,8 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
 	config := clientcmdapi.NewConfig()
-	mergo.Merge(config, mapConfig)
-	mergo.Merge(config, nonMapConfig)
+	mergo.MergeWithOverwrite(config, mapConfig)
```
Previously, map values were only overwritten if no key was set in the destination map before.  With this change, the destination map value will be overwritten.  I think the new semantic could be used to make this function easier to ready by eliminating the distinction between `mapConfig` and `nonMapConfig`, but as it is now isn't the `mapConfig` essentially stomped?

peter-edge replied:
This isn't completely accurate - previously, values would be stomped if a new non-zero value was set in object being merged. That this changed to no overwrite is what caused https://github.com/GoogleCloudPlatform/kubernetes/issues/12272.

I can look at this specific code in a bit but the intention here was just to maintain existing functionality.

deads2k replied:
> This isn't completely accurate - previously, values would be stomped if a new non-zero value was set in object being merged. That this changed to no overwrite is what caused #12272.

Map values were first write wins, not last write wins.  Struct values were last write wins.  `MergeWithOverwrite` makes the map values last write wins too.  Hence all the comments explaining why multiple objects were merged in reverse orders.

peter-edge replied:
ah noted. ok I will check this all out in a bit and update, sorry bout that

deads2k commented on pkg/client/clientcmd/client_config.go:
```diff
@@ -109,13 +109,13 @@ func (config DirectClientConfig) ClientConfig() (*client.Config, error) {
 			return nil, err
 		}
-		mergo.Merge(clientConfig, userAuthPartialConfig)
+		mergo.MergeWithOverwrite(clientConfig, userAuthPartialConfig)
```
You'll need to update the comments a couple lines up.  They refer to the old merge semantics.

peter-edge replied:
will do!

peter-edge commented:
@deads2k another option i hate to mention is we have kubernetes' own version of mergo in pkg/ and not rely on an external library. for something like this, it might not be a bad idea.

deads2k commented:
> @deads2k another option i hate to mention is we have kubernetes' own version of mergo in pkg/ and not rely on an external library. for something like this, it might not be a bad idea.

I don't see anything particularly wrong with using this external library.  I'm just pretty sure that semantics changed (for the better I think) and we'll update with it.

PhilibertDugas committed:
Tuned the kube-proxy upstart script to be in line with:
https://github.com/GoogleCloudPlatform/kubernetes/issues/5461

Upstart scripts need to specify explicitely the nofile limit

smarterclayton committed:
Extra newlines in error output

StandardErrorMessage does not have a newline by default, other
error messages should not end with a newline.

gmarek committed:
Merge pull request #12453 from epc/epc-2015221

Copy edits for typos

gmarek committed:
Merge pull request #12448 from smarterclayton/extra_newlines

Extra newlines in error output

gmarek committed:
Merge pull request #12443 from PhilibertDugas/master

Tuned the kube-proxy upstart script

gmarek committed:
Merge pull request #11788 from HaiyangDING/ImproveClarityResource

Improve clarity around PodFitsResource(issue#11453)

bufdev committed:
Merge branch 'mergo-update' of https://github.com/peter-edge/kubernetes into mergo-update

peter-edge commented:
Closing this for now and will redo in another PR - had a git snafu locally and will be easier to just do the substitute/replace in a new PR.
